### ISSUE TITLE
Support for Internalizing/Merging external Dependencies

### ISSUE THREAD

**Issue by avanderhoorn** [aspnet/dnx#819]
## tl;dr
In a world where an ever growing number of packages are depending on other packages for "internal" functionality, the amount of unresolvable conflicts that occur between packages will increase. This specification is designed to help package authors to explicitly declare these internal dependencies so that the runtime or build process can allow multiple versions of the same package to run, and so that users aren't faced with unresolvable conflicts.

### Internal Dependency
An Internal dependency is defined as a dependency that a controlling package needs for its own internal functionality. To uses/consumers of the controlling package, they should not be affected or constrained by which version of a package the controlling package chooses to use. An internal dependency should not "leak" outside of the controlling package and should be able to operate in its own memory space (meaning if 2 versions of the same package are present, there should not be any conflict between the 2 packages).

**Example**
Glimpse is a tool based package which users add to their system to gain insights on how their system is running. INTERNALY we are using JSON.Net to seralize data for our own internal consumption. The JSON.Net types are never leaked and the users are never aware that we use or what version we are running. To the user we might as well have written our own seralizer.

### Public Dependency
A Public dependency is defined as a dependency that an authors package extends and/or exposes the dependent packages types. To uses/consumers of the authors package, they need to be aware of the version of a package the controlling package chooses to use as it adds/extends a package (at a specific version) that they are using. An public dependency most likely has types that "leak" outside of the authors package and should be able to operate in the same memory space (as it needs to interact with types and memory in that public dependency).

**Example**
LinqKit is a set of extensions that is built on top of Entity Framework. It needs to publicly depend on Entity Framework as it is extending EF types/functionality and is dependent on the version of EF that the user is using. 

### Summary
When you are extending something, you don't internalize (as that obviously doesn't work) but if you depend on something just for internal use, Internal is used to avoid issues

## Overall Problem

In a world were package authoring and consumption is becoming ever easier, it should be expected that the number of conflicts between packages will only increase. In this world there there are Public dependencies which MUST be resolved in order to function as expected and Internal dependencies which shouldn't effect the package version choices of someone using the internal dependences. Moving forward, the amount of packages that have truly internal dependencies will most likely increase and the pain that authors and consumers experience increase as a result.  

**Version conflicts will be a problem**
In my app, if I use 5 - 7 different packages for middleware functionality, its highly unlikely that all these will use the same version of JSON.Net or log4net. In fact it is more likely that they will want to use 5 - 7 different versions of JSON.Net or log4net. Expecting the user to resolve and *test* these conflicts is a high expectation and they mightn't even be able to resolve those conflicts (as they might all be using different incomparable versions). And asking package authors to change there versions wont really work either, as no change will work for everyone.

Internal dependencies are designed to address these issues and allow package authors to be explicit about there Internal dependencies. 

## Specification

**Not leaking types**
*When a package author defines a dependency as being Internal, none of the types from that Internal dependency should leak through the public API.*

This limitation simplifies the problem space and reflects a constraint which most users internalizing dependencies via ILMerge (or other similar methods are used to) are used to. When types can't be leaked it means that we need to build abstractions to let users toggle given configurations or proxy given given providers, but that "limitation" also acts as a point of clear separation between the my code and the package that I (as a package author) take a dependency on (like JSON.net or log4net, etc). It also means that a package author can update internal dependencies without having to effecting their users.

**Dependent packages operate in their own memory space**
*When a package author defines a dependency as being Internal, any state or memory that the internal package sets, should not conflict with any other instance of that package that is running.*

Depending on how the feature is implemented, this might come as a result, but if a package that I depend on has a static variable, I don't want this to conflict with the same package that the user may be depending on directly. This might also come as a side effect of not leaking types (if that is implemented by changing things to internal) or might have to be done by changing namespaces or some other method, but it would be expected that dependent packages operate in their own memory space.

**Support in the platform**
*When a package author defines a dependency as being Internal, it should be supported by the platform so that authors and consumers alike have a consistent and supported experience*

Having support in the platform for this problem means that instead of everyone going off and trying to solve it in different ways, we are combining our collective intelligence to have a process that is provided out of the box. It also means that consumer of package who use this, have a common expectation of how things will work.

## Guidlines and Expectations

**When a author internalizes a package, it now "owns" that dependency**
If my package depends on JSON.Net and I declare that dependency as Internal, I am now taking responsibility for that code. For consumers of my package, they should no longer think of my package as depending on JSON.Net, but just my package. They probably won't even know that have a dependency on JSON.Net. As far as they are concerned, I might as well built my own JSON seralizer. This is consistent with the fact that Types aren't leaked.

**Package author controlling the version**
I've seen it said that "what if the package user wants to change the version of JSON.Net/log4net that we are using in Glimpse". I struggle to understand what the requirement is here. If we use version 1.1 of log4net and a user try's to change it to 1.2, things will most likely not work correctly. When we currently ILMerge something in, that is us (the package author) taking responsibility for that dependent package. How can we make an guarantees about our code running correctly, if people have the ability to change the version of one of our dependencies without us having tested that. If a bug is found in a package that we depend on, it is up to us to rerelease our package or the user to get the source and rebuild it. Again trying to make simple rules. Semver is great, but it doesn't guarantee that expected APIs or behaviours wont change/break.

**Don't have to support every use case, this isn't a silver bullet**
I think that its ok if support for this feature doesn't work in absolutely every use case. For instance, Castle requires that some types **must** be public and hence can't be fully Internalized, that is ok. If we end up supporting the use case where we can change namespaces and log4net is doing string comparison on its types, pressure will soon come from the community to change it. In the end, with power like this also comes great responsibility. Its up to the author to make sure things work. 

## Implementation
At this point, this specification isn't worried about how this could be implemented. This is intended to document a use case and state what the base expectations are, we can work on implementation details later.

## Example

Current thinking is that package authors could express internal dependency via the `project.json`. This being the case, changes to the `project.json` could look as follows
```
"aspnet50": {
  "dependencies": {
    "EntityFramework": "6.1.2-beta1",
    "NLog": { "version": "1.0.0.505", "type": "internal" },
    "Newtonsoft.Json": { "version": "6.0.1", "type": "internal" }
  }
},
```

In the above example, the package has a Public dependency on Entity Framework, but Internally depends on NLog and JSON.Net for its internal logging and serialization.

---------------

## Original Comment

**Ideally/Long Term:**
Currently, dependencies can we setup for external packages:
```
"aspnet50": {
  "dependencies": {
    "Microsoft.Framework.Logging.Interfaces": "1.0.0-*"
  }
},
```

These dependencies can also be marked as build dependencies only:
```
"aspnet50": {
  "dependencies": {
    "Microsoft.Framework.Logging.Interfaces": { "version": "1.0.0-*", "type": "build" }
  }
},
``` 

It would be nice if one could mark the dependency as being "internal":
```
"aspnet50": {
  "dependencies": {
    "Microsoft.Framework.Logging.Interfaces": { "version": "1.0.0-*", "type": "internal" }
  }
},
``` 

In the end the definition might be different to the above, but the end result is that the dependency would be merged into the "main" dll. 

**Hack/Short Term**
In the mean time, it would be great if there was a support way for us to be able to do this (probably with a [script](https://github.com/aspnet/Home/wiki/Project.json-file#scripts)) so that it works with the existing `kpack` process. 

If the correct hook is already in place, can you please point me in the right direction for this "Hack/Short Term" solution.

*This is an extension of a discussion that I had with @davidfowl on JabbR.*

**Comment by jchannon** [comment]
If we could have a node like dependency story for .Net that'd be :+1:

**Comment by McDonnellDean** [comment]
+1 to the updated proposed spec!

**Comment by ma8anm** [comment]
+1 from me. I have lost a lot of time to these sorts of problems in the past, to the point of splitting deployments etc to get around it which is nuts!

**Comment by ChrisMcKee** [comment]
Great idea :+1:

**Comment by adamralph** [comment]
@jchannon https://github.com/aspnet/KRuntime/issues/819#issuecomment-62609808 and https://github.com/aspnet/KRuntime/issues/819#issuecomment-62709041.

Take it to the extreme by having source only. I still think it could work, but it's a radical departure from where things are now.

**Comment by forki** [comment]
(@adamralph Paket is very very slowly going in this direction http://fsprojects.github.io/Paket/github-dependencies.html#Remote-dependencies)

**Comment by adamralph** [comment]
@forki Nice! How does it handle namespaces? And does it internalise?

**Comment by forki** [comment]
at the moment this feature is "just" downloading and adding source files to the project and it retrieves indirect dependencies as well.

**Comment by damianh** [comment]
I'd prefer static linking over a node like dependency story.

**Comment by damianh** [comment]
Had a wonderful argument with TinyIoC with someone because of style cop
rules, source analysis and generally not being consistent with the rest of
the code base :(

**Comment by adamralph** [comment]
@damianh see https://github.com/aspnet/KRuntime/issues/819#issuecomment-62709041

Under my proposal, the packaged source files would not be considered part of the codebase for things like static analysis. Of course, they would have to be compiled, but in most, of not all, other contexts they would be treated as being something external.

**Comment by damianh** [comment]
Lads, see the recent C# Design Notes https://github.com/dotnet/roslyn/issues/98 , under Componentization - "static linking instead of IL merge"

I think we're getting somewhere, I'm holding my breath :D

**Comment by davkean** [comment]
@damianh http://blogs.msdn.com/b/dotnet/archive/2014/04/02/announcing-net-native-preview.aspx

**Comment by damianh** [comment]
@davkean Correct me if I'm wrong, but this is more for end apps (.exes) and not libs, right?

**Comment by ayende** [comment]
That would be great for libs, if we can do that.

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by damianh** [comment]
https://github.com/dotnet/roslyn/issues/162

**Comment by davidfowl** [comment]
Just a small update on this. We're messing around with internalizing ourselves (very manually) via roslyn magic and git submodules https://github.com/aspnet/XRE/pull/1239/files

**Comment by avanderhoorn** [comment]
@davidfowl O-O... WOW, talk about force of will. Kinda cool that the XRE "lets" you do things like this. Assuming its proved out a little more, this could totally be made into something a little more supported. At the very least it could be a package that acts as a helper in doing things like this... save everyone writing the same code.

**Comment by quintonn** [comment]
Would it not be a good idea to treat different versions of the same library as completely different libraries? I think this is what @ayende was referring to?
I.E, if using log4net v 1.2.10, then it would add a reference to a dll called log4net.1.2.10.dll, which will not be the same as log4net.1.2.11.dll. Or perhaps better would be to prefix the dll's with the name of the package it came with. 
Eg. My produc ABC requires log4net, so when i compile and package my product there will be a dll called "abc_log4net.dll". But i think the first option is better.
This would mean, however, that passing objects from one version of a library to another would require recreating objects or remapping properties, but at least that is something that can be done in code and checked at compile time instead of at run time.
I have used ILMerge a bit and it's not easy. Last time i checked it couldn't merge a WPF application and all it's dll's (at all).

**Comment by davidfowl** [comment]
This was a great discussion and should be brought up again on either CoreFX or Roslyn or both.
